Src/FluentAssertions/Primitives/TimesConstraint.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using FluentAssertions.Common;
+using FluentAssertions.Execution;
+
+namespace FluentAssertions.Primitives
+{
+    public abstract class TimesConstraint
+    {
+        private int? actualCount;
+
+        protected readonly int expectedCount;
+
+        public TimesConstraint(int expectedCount)
+        {
+            this.expectedCount = expectedCount;
+        }
+
+        protected int ActualCount
+        {
+            get
+            {
+                if (!actualCount.HasValue)
+                {
+                    actualCount = Subject.CountSubstring(Expected, StringComparison);
+                }
+
+                return actualCount.Value;
+            }
+        }
+
+        protected abstract string Mode { get; }
+
+        protected abstract bool IsMatch { get; }
+
+        private string Expected { get; set; }
+
+        private string Subject { get; set; }
+
+        private StringComparison StringComparison { get; set; }
+
+        private static string Times(int count) => count == 1 ? "1 time" : $"{count} times";
+
+        internal void AssertContain(string subject, string expected, string because, params object[] becauseArgs)
```

Comment:
Since this class knows nothing about the algorithm that you use (_contains_), I think just naming it `Assert` would be enough. This also opens up the path for reuse (even though we would have to solve the hard-coded string in `FailWith` )

Reply:
I see where you're going, but the current implementation is specifically targeting strings. What shall I do with the other `internal` method `AssertContainEquivalentOf`? That one wouldn't make sense for something like an `int[]`. This is where the re-usability of the current implementation is lacking in my opinion. Maybe `OccuranceConstraint` should be made generic? I'm not sure the best way to go about this.

Reply:
True. So maybe it's fine if we keep it all locked down as much as possible until we find a proper we to reuse it.

Reply:
@dennisdoomen I've dumbed down the `OccurrenceConstraint` type, so that it only knows about the expected count, actual count, and the relationship between them, in the form of an `Assert` method. This way the responsibility of counting and producing an assertion message is given back to the FA assertion method, and `OccurrenceConstraint` can safely be reused in the public API.

Can you have a look and tell me what you think? 👍

Reply:
Looks good enough for me.